src/rt-app.c

```diff
@@ -794,26 +803,44 @@ void *thread_body(void *arg)
 			log_timing(data->log_handler, curr_timing);
 
 		if (opts.ftrace)
-			log_ftrace(ft_data.marker_fd, "[%d] end loop %d phase %d step %d",
-				   data->ind, i, j, loop);
-
-		loop++;
-		if (loop == pdata->loop) {
-			loop = 0;
-
-			j++;
-			if (j == data->nphases) {
-				j = 0;
-				i++;
+			log_ftrace(ft_data.marker_fd, "[%d] end loop %d phase %d",
+				   data->ind, phase_loop, phase);
+
+		phase_loop++;
+		/* Reached the specified number of loops for this phase. */
+		if (phase_loop == pdata->loop) {
+			/* Check if we have a following phase to execute */
+			phase++;
+			if (phase != data->nphases) {
+				/*
+				 * Yes, we have it, so execute it
+				 * (and reset loop count for it)
+				 */
+				phase_loop = 0;
+				pdata = &data->phases[phase];
+			} else if (data->loop == -1 || thread_loop < data->loop - 1) {
+				/*
+				 * Phases are finished, but we might
+				 *
+				 * - have to execute indefinetly
+				 *   (data->loop, i.e. task level "loop", == -1)
+				 * - perform a number of task level iterations of
+				 *   the same phase(s)
+				 *
+				 * In both cases, start over from first phase.
+				 *
+				 */
+				phase_loop = 0;
+				phase = 0;
+				thread_loop++;
+				pdata = &data->phases[0];
```

Comment:
This seems to be quite complex compared to what was done before. IIUC, you don't use thread_loop anymore in the while because it will overflow and equals -1 at some point and the thread will stop, ins't it ?
I this case why not doing:
```
		phase_loop++;
		/* Reached the specified number of loops for this phase. */
		if (phase_loop == pdata->loop) {
			phase_loop = 0;

			phase++;
			/* Finished the last phase */
			if (phase = data->nphases) {
				phase = 0;

				thread_loop++;
				/* 
				 * a negative thread_loop means that it has overflowed and
				 * that data->loop == -1 which means infinite loop.
				 * Reset thread_loop to prevent stopping the thread
				 */
				if (thread_loop < 0)
					thread_loop = 0;
			}
			pdata = &data->phases[phase]
		}
```